src/bit-fields-macros/src/lib.rs

```diff
@@ -0,0 +1,1070 @@
+             `u128`."
+        )
+    }
+}
+impl std::error::Error for DataTypeTryFromError {}
+
+impl TryFrom<&str> for DataType {
+    type Error = DataTypeTryFromError;
+    fn try_from(s: &str) -> Result<Self, Self::Error> {
+        match s {
+            "u8" => Ok(Self::U8),
+            "u16" => Ok(Self::U16),
+            "u32" => Ok(Self::U32),
+            "u64" => Ok(Self::U64),
+            "u128" => Ok(Self::U128),
+            _ => Err(DataTypeTryFromError),
+        }
+    }
+}
+
+/// Utility used to assembly the display string
+#[derive(Debug)]
+struct MultiLineString(pub Vec<String>);
+impl MultiLineString {
+    /// Pushes `MultiLineString` on to `self`, if the given `other` is deeper than `self` it will
+    /// expand down to accommodate.
+    // None of these indexes or arithmetic operations will ever panic, thus we can safely use
+    // unchecked operations here. Since unchecked arithmetic is currently unstable we just allow
+    // integer arithmetic for now.
+    #[allow(clippy::integer_arithmetic)]
+    fn push(&mut self, other: &Self) {
+        unsafe {
+            let mut i = 0;
+            while i < std::cmp::min(self.0.len(), other.0.len()) {
+                self.0
+                    .get_unchecked_mut(i)
+                    .push_str(other.0.get_unchecked(i));
+                i += 1;
+            }
+            while i < other.0.len() {
+                self.0.push(other.0.get_unchecked(i).clone());
+                i += 1;
+            }
+        }
+    }
+    /// Pushes a static string ont `self`.
+    fn push_str(&mut self, s: &str) {
+        self.push(&MultiLineString::from(s));
+    }
+}
+impl From<&str> for MultiLineString {
+    fn from(s: &str) -> Self {
+        Self(s.split('\n').map(String::from).collect())
+    }
+}
+impl From<MultiLineString> for String {
+    fn from(mls: MultiLineString) -> Self {
+        let mut iter = mls.0.into_iter();
+        if let Some(mut base) = iter.next() {
+            for next in iter {
+                base.push('\n');
+                base.push_str(&next);
+            }
+            base
+        } else {
+            String::new()
+        }
+    }
+}
+impl fmt::Display for MultiLineString {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for line in &self.0 {
+            writeln!(f, "{}", line)?;
+        }
+        Ok(())
+    }
+}
+
+// TODO Allow writing rustdoc comments on bitfield structs
+/// Convenience macro for errors in [`bitfields`].
+macro_rules! macro_error {
+    ($span:expr,$msg:expr) => {{
+        proc_macro_error::Diagnostic::spanned($span.span(), Level::Error, $msg.into()).emit();
+        return proc_macro::TokenStream::new();
+    }};
+}
+
+/// Procedural macro to generate bit fields.
+///
+/// ```ignore
+/// #[rustfmt::skip]
+/// bit_fields::bitfield!(GeneratedBitField,u32,[
+///     RANGE1, 0..1,
+///     SSE, 2,
+///     SSE1 3,
```

Comment:
I tried testing this but failed (`SSE1, 3`)

Reply:
The generated code requires structures from the `bit-fields` crate, this is why this test is ignored and cannot run.